### PR TITLE
Rewrite Primer Spec using Preact

### PR DESCRIPTION
**Pull request by seshrs**
## TLDR

This PR proposes a major refactor of the Primer Spec JS, centered around a switch to using [Preact](https://preactjs.com/) (a React alternative).

The PR is monstrously large, and I don't expect a thorough review of the code changed in this PR. However, I'd really appreciate if people could use various browsers/devices to visit the validation URL at https://preview.seshrs.ml/previews/eecs485staff/primer-spec/63/ and play around with the page. (Please comment if you do so!)

#### Why refactor?
Primer Spec started off as inline JavaScript + JQuery embedded in `spec.html`. As I started adding more features (especially subthemes and mobile-friendliness), I figured Primer Spec was a complex-enough project that modularized JS made sense. At the time, I felt it might be over-kill to use the "heavy" React framework (see #13). (In retrospect, I think it was more because I couldn't envision what a React-ized Primer Spec would look like.)

Recently, while trying to add even more customizability to Primer Spec (including options to disable sidebars/subthemes, set a custom default subtheme, etc.), I felt that implementing those features would be complex, and involve spaghetti-like dependencies across components. Over the past few months, I decided that switching to React would be the best longer-term solution.

#### Why [Preact](https://preactjs.com/)?
Because it's _almost_ React, and much smaller! According to [Bundlephobia](https://bundlephobia.com/), React + ReactDom adds ~120kB to the bundle size, whereas Preact adds only 10kB.

React has more features and optimizations than Preact, and works better for large complex apps. For now, I feel Preact is sufficient for Primer Spec. Also, migrating from Preact to React in future should be pretty simple.

#### Is Primer Spec slower now?
No, Primer Spec actually became faster *and* smaller with this refactor.

- **Bundle size:** Reduced from 120kB to 39kB (67% smaller!)
  - This is mainly because I removed JQuery and unused JS dependencies from the upstream Primer repo.
- **Load time:** Approximately similar, but "feels" faster. This is because there are no layout shifts! (The Sidebar is fully rendered when the page "loads".)
- **Script overload:** It's difficult to perceive the difference on modern computers. But when I simulated a 6x CPU throttle (using Chrome DevTools), the refactor used 50% less CPU time than before.

#### How is Primer Spec JS now organized?
Here's an overview. [(Source)](https://excalidraw.com/#json=5175969081982976,eIOihS_d01XGb99ft-L-OA)

Primer Spec (Preact) Overview

#### What about state-management frameworks like Redux?
I was actually considering using [Redux-Zero](https://github.com/redux-zero/redux-zero), but it was simple enough to maintain all state at the top-level parent and pass them down as props to children. I don't anticipate Primer Spec getting much more complex, so I don't think we need a state-management framework for now.

#### What else changed?
In addition to refactoring the entire JS functionality, I included the following updates:
- If you resize the window's width to under 900px, "mobile mode" gets activated automatically. The reverse is also implemented.
- If you scroll to the top of the page and no sidebar items are "active", then none are highlighted. (Previously, the first sidebar item would always be highlighted even if you scroll to a previous section.) The demo page at the [validation URL](https://preview.seshrs.ml/previews/eecs485staff/primer-spec/63/) actually demonstrates this behavior.
- The docs now recommend enabling certain Jekyll configurations that are usually used by default on GitHub Pages, including GitHub Flavored MarkDown (GFM), converting `README.md` to `index.md`, and using `_layouts/default.html` to render all MarkDown files by default.

#### Next Steps?
I'd like to add some per-page and per-site configuration options for Primer Spec for the upcoming Fall Semester (tracked by the [Fall 2020 Milestone](https://github.com/eecs485staff/primer-spec/milestone/2)).

In the near-term, I'd also like to explore options for including CSS-in-JS. The styles are quite complex and messy, with tons of class-names all over the Preact code 😛 

## Validation

Please visit https://preview.seshrs.ml/previews/eecs485staff/primer-spec/63/ on various browsers and devices to make sure the page works as expected. Let me know if anything seems off. Thanks!

CC @awdeorio, and volunteers from EECS 485 staff

### CODE REVIEW

**Comment by seshrs** [comment]
I just noticed 2 issues on iOS Safari:
- Sidebar doesn’t close when I click outside it. 
- LaTeX is not rendered. 

Will address these issues soon.

File: README.md [hunk 1]
```diff
@@ -3,47 +3,59 @@
...
 
 Primer Spec is built on top of the wonderful [Primer theme](https://github.com/pages-themes/primer), and adds functionality useful for pages with a lot of content. This theme was primarily designed for hosting project specifications for EECS courses at the University of Michigan. See the [User Showcase](#user-showcase) for inspiration.
 
 ## Contents
-- [Usage](#usage)
-- [User Showcase](#user-showcase)
-- [Contributing](#contributing)
-- [Acknowledgements](#acknowledgements)
+
+- [The Primer Spec theme](#the-primer-spec-theme)
```
Comment (seshrs) [inline README.md:17]: A VSCode plugin made this change. Will revert.

File: webpack.config.js [hunk 2]
```diff
@@ -95,17 +95,14 @@ module.exports = (env) => ({
   },
   // When importing files, no need to mention these file-extensions
   resolve: {
-    extensions: ['.js', '.ts'],
+    extensions: ['.js', '.ts', '.tsx'],
   },
   plugins: [
-    // JQuery becomes available in every file
```
Comment (seshrs) [inline webpack.config.js:101]: I should double-check that I removed references to the liquid-loader.

File: src_js/components/sidebar/TableOfContents.tsx [hunk 3]
```diff
@@ -0,0 +1,126 @@
...
+    const hash_listener = () => {
+      if (props.isSmallScreen && props.sidebarShown) {
+        props.onToggleSidebar();
+      }
+      if (props.settingsShown) {
+        props.onToggleSettings();
+      }
+    };
+
+    window.addEventListener('hashchange', hash_listener, false);
```
Comment (seshrs) [inline src_js/components/sidebar/TableOfContents.tsx:42]: One issue with listening to `hashchange` is that if the user clicks the same sidebar item consecutively, subsequent clicks do not trigger this event. (So on mobile, if a user clicks a section, scrolls, then opens the sidebar to click the same section, the sidebar won't close.)

I think this is an acceptable bug for now, but it should be documented by opening an issue.

File: src_js/components/sidebar/TableOfContents.tsx [hunk 4]
```diff
@@ -0,0 +1,126 @@
...
+ * Previous versions of Primer Spec used a custom fork of
+ * htmlTableOfContents.js by Matthew Christopher Kastor-Inare III
+ * (https://github.com/matthewkastor/html-table-of-contents).
+ * The current implementation bears very little resemblance to the original
+ * algorithm besides the structure of the final HTML output.
+ *
+ * @param headings List of HTML nodes representing header elements in the page
+ * @param activeHeadingIndex The index of the active heading item. Use -1 to deactivate.
+ */
+function generateTocNodes(headings: Element[], activeHeadingIndex: number) {
```
Comment (seshrs) [inline src_js/components/sidebar/TableOfContents.tsx:94]: Maybe move this method to a separate module?

Comment (seshrs) [reply src_js/components/sidebar/TableOfContents.tsx:94]: Actually, I'll consider moving this out in future. (It's a core part of the TableOfContents logic, so I think I'll leave it there for now.)

File: jekyll-theme-primer-spec.gemspec [hunk 5]
```diff
@@ -21,4 +21,6 @@ Gem::Specification.new do |s|
   s.add_development_dependency 'html-proofer', '~> 3.0'
   s.add_development_dependency 'rubocop', '~> 0.50'
   s.add_development_dependency 'w3c_validators', '~> 1.3'
+
```
[the remark this answers is not in the document]
Comment (seshrs) [reply jekyll-theme-primer-spec.gemspec:24]: Turns out this won't work with remote-theme. :(
